File: clients/android/app/src/main/java/com/audiobook/app/AuthViewModel.kt
package com.audiobook.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.audiobook.core.domain.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class AuthUiState(
    val email: String = "",
    val password: String = "",
    val isLoading: Boolean = false,
    val isAuthenticated: Boolean = false,
    val errorMessage: String? = null
)

class AuthViewModel(private val repository: AuthRepository) : ViewModel() {
    private val _state = MutableStateFlow(AuthUiState(isAuthenticated = repository.isAuthenticated()))
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    fun onEmailChanged(value: String) {
        _state.value = _state.value.copy(email = value)
    }

    fun onPasswordChanged(value: String) {
        _state.value = _state.value.copy(password = value)
    }

    fun login() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, errorMessage = null)
            try {
                repository.login(_state.value.email, _state.value.password)
                _state.value = _state.value.copy(isLoading = false, isAuthenticated = true)
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    isAuthenticated = false,
                    errorMessage = "Login failed. Check credentials and try again."
                )
            }
        }
    }

    fun refreshSession() {
        viewModelScope.launch {
            try {
                repository.refreshSession()
                _state.value = _state.value.copy(isAuthenticated = true, errorMessage = null)
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isAuthenticated = false,
                    errorMessage = "Session expired. Please sign in again."
                )
            }
        }
    }

    fun signOut() {
        repository.signOut()
        _state.value = _state.value.copy(isAuthenticated = false, password = "")
    }
}

File: clients/apple/AudiobookAppleApp/Features/Auth/ViewModels/AuthViewModel.swift
import Foundation
import AudiobookCore

@MainActor
final class AuthViewModel: ObservableObject {
    @Published private(set) var state = AuthViewState()

    private let authService: AuthService

    init(authService: AuthService) {
        self.authService = authService
        self.state.isAuthenticated = authService.isAuthenticated
    }

    func updateEmail(_ value: String) {
        state.email = value
    }

    func updatePassword(_ value: String) {
        state.password = value
    }

    func login() async {
        state.isLoading = true
        state.errorMessage = nil

        do {
            try await authService.login(email: state.email, password: state.password)
            state.isAuthenticated = true
        } catch {
            state.errorMessage = "Login failed. Check credentials and try again."
            state.isAuthenticated = false
        }

        state.isLoading = false
    }

    func refreshSession() async {
        do {
            try await authService.refreshSession()
            state.isAuthenticated = true
        } catch {
            state.errorMessage = "Session expired. Please sign in again."
            state.isAuthenticated = false
        }
    }

    func signOut() {
        authService.signOut()
        state.isAuthenticated = false
        state.password = ""
    }
}
